### PR TITLE
librbd: explicitly specify mode on mirror image enable

### PR DESCRIPTION
**Pull request by trociny**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/librbd/api/Mirror.cc [hunk 1]
```diff
@@ -392,8 +395,20 @@ int Mirror<I>::image_enable(I *ictx, bool relax_same_pool_parent_check) {
     }
   }
 
+  if (!ictx->test_features(RBD_FEATURE_JOURNALING)) {
```
Comment (dillaman) [inline src/librbd/api/Mirror.cc]: Nit: only for `mode == MIRROR_IMAGE_MODE_JOURNAL`

File: src/librbd/api/Mirror.cc [hunk 2]
```diff
@@ -529,6 +544,12 @@ int Mirror<I>::image_disable(I *ictx, bool force) {
       rollback = true;
       return r;
     }
+
+    r = ictx->operations->update_features(RBD_FEATURE_JOURNALING, false);
```
Comment (dillaman) [inline src/librbd/api/Mirror.cc]: Nit: only for `mode == MIRROR_IMAGE_MODE_JOURNAL`

File: src/librbd/api/Trash.cc [hunk 3]
```diff
@@ -84,6 +84,10 @@ int enable_mirroring(IoCtx &io_ctx, const std::string &image_id) {
     return r;
   }
 
+  if ((features & RBD_FEATURE_JOURNALING) == 0) {
```
Comment (dillaman) [inline src/librbd/api/Trash.cc]: Hmm ... so we don't have state available to restore the original mirror mode if you move it to the trash?

File: src/librbd/operation/EnableFeaturesRequest.cc [hunk 4]
```diff
@@ -156,67 +159,7 @@ Context *EnableFeaturesRequest<I>::handle_get_mirror_mode(int *result) {
...
-    return handle_finish(*result);
-  }
-
-  if (mirror_image.mode == cls::rbd::MIRROR_IMAGE_MODE_SNAPSHOT) {
-    lderr(cct) << "cannot enable journaling: image snapshot mirroring enabled"
-               << dendl;
-    *result = -EINVAL;
-    return handle_finish(*result);
-  }
+  m_enable_mirroring = (mirror_mode == cls::rbd::MIRROR_MODE_POOL);
```
Comment (dillaman) [inline src/librbd/operation/EnableFeaturesRequest.cc]: Nit: maybe here we could use a `rbd_default_mirror_image_mode` config override

File: src/tools/rbd/action/MirrorImage.cc [hunk 5]
```diff
@@ -104,7 +128,13 @@ int execute_enable_disable(const po::variables_map &vm, bool enable,
     return r;
   }
 
-  r = enable ? image.mirror_image_enable() : image.mirror_image_disable(force);
+  librbd::mirror_image_mode_t mode = RBD_MIRROR_IMAGE_MODE_JOURNAL;
+  if (vm.count("mode")) {
```
Comment (dillaman) [inline src/tools/rbd/action/MirrorImage.cc]: Does this work for positionals?

File: src/librbd/api/Trash.cc (hunk 3 above)
Comment (trociny) [reply src/librbd/api/Trash.cc]: We enable mirroring here only if the pool in mirror mode "pool", which we decided should only be used for journal mirroring.

File: src/librbd/operation/EnableFeaturesRequest.cc (hunk 4 above)
Comment (trociny) [reply src/librbd/operation/EnableFeaturesRequest.cc]: I am not sure I understand why we need `rbd_default_mirror_image_mode` here. When enabling journaling feature, we want to enable mirroring only when the pool in `MIRROR_MODE_POOL`, and it looks ok to me?

File: src/tools/rbd/action/MirrorImage.cc (hunk 5 above)
Comment (trociny) [reply src/tools/rbd/action/MirrorImage.cc]: Indeed. Thanks. We don't have a test  to catch this right now yet. I suppose we will have one in qa/workunits/rbd/rbd_mirror.sh when rbd-mirror supports snapshot mirroring.

File: src/librbd/api/Trash.cc (hunk 3 above)
Comment (dillaman) [reply src/librbd/api/Trash.cc]: Ack -- thx

File: src/librbd/operation/EnableFeaturesRequest.cc (hunk 4 above)
Comment (dillaman) [reply src/librbd/operation/EnableFeaturesRequest.cc]: I forgot about that idea that pool-mode implicit enable/disable would only work for journal-based mirroring. Disregard.

**Comment by trociny** [comment]
@dillaman updated. I think I addressed all your comments except those where I have questions.

Also added some small improvements to `rbd`.

**Comment by dillaman** [comment]
I haven't looked into it yet to determine the true cause (three rbd-mirror PRs in one branch), but the rbd-mirror workunit failed:

http://qa-proxy.ceph.com/teuthology/jdillaman-2019-12-15_08:02:18-rbd-wip-jd-testing-distro-basic-smithi/4604979/teuthology.log

**Comment by dillaman** [comment]
... looks like in the `TEST: disable mirroring / delete non-primary image` test, the mirror pool mode is switched from `pool` to `image`. Mirroring is disabled on images `test2` and `test4` which turns off the `journaling` feature, so when the mirror pool mode is switched back to `pool`, those two images are not mirrored as expected.

**Comment by trociny** [comment]
@dillaman ah, thank you. I forgot to revert the changes to rbd_mirror.sh I did for my initial "implicit mode" implementation, and somehow overlooked the test was failing. I have reverted it back.

**Pull request merged**